File: CCVTAC/CCVTAC.Console/PostProcessing/Renamer.cs
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace CCVTAC.Console.PostProcessing;

internal static class Renamer
{
    private readonly record struct RenamePattern(Regex Regex, string ReplaceWithPattern, string? Description = null);

    // TODO: Convert this into a setting.
    private static readonly IReadOnlyList<RenamePattern> RenamePatterns = new List<RenamePattern>()
    {
        // Universal, to always be run first.
        new(
            new Regex(@"\s\[[\w_-]{11}\](?=\.\w{3,5})"),
            string.Empty,
            "Remove trailing video IDs (recommend running this first)"),
        new(
            new Regex(@"\s{2,}"),
            " ",
            "Remove extra spaces"),
        // Various patterns
        new(
            new Regex(@"(?<= - )\d{3} (\d{1,3})\.?\s?"),
            "%<1>s - ",
            "Remove and reformat duplicate track numbers"),
        new(
            new Regex(@"\s*[(（【［\[\-]?(?:[Oo]fficial +|OFFICIAL +)?(?:HD )?(?:[Mm]usic [Vv]ideo|MUSIC VIDEO|[Ll]yric [Vv]ideo|LYRIC VIDEO|[Vv]ideo|VIDEO|[Aa]udio|[Vv]isualizer|AUDIO|[Ff]ull (?:[Aa]lbum|LP|EP)|M(?:[_/])?V)[)】］）\]\-]?"),
            string.Empty,
            "Remove unneeded labels"),
        new(
            new Regex("""\s?[\(（]歌詞入り[\)）]"""),
            string.Empty,
            "Remove \" (歌詞入り)\""),
        new(
            new Regex("""【(.+)】(.+)"""), // 【person】title
            "%<1>s - %<2>s",
            "PERSON - TRACK"),
        new(
            new Regex(@"(.+?)(?: - )(.+?) \[[\w⧸]+\] .+ \(([\d\?？]{4})\)"),
            "%<1>s - %<2>s [%<3>s]",
            "PERSON - TRACK [YEAR]"),
        new(
            new Regex(@"^(.+?)(?: - )?\s?[｢「『【](.+)[」｣』】]\s?\[?([12]\d{3})\]?(?:\s?MV)?"),
            "%<1>s - %<2>s [%<3>s]",
            "Reformat 'PERSON「TITLE」YEAR' and 'PERSON「TITLE」[YEAR]'"),
        new(
            new Regex(@"^(.+?)(?: - )?\s?[｢「『【](.+?)[」｣』】](?:\s?MV)?(?=\.\w{3,4})"),
            "%<1>s - %<2>s",
            "Reformat 'PERSON「TITLE」' (alone, not followed by anything)"),
        new(
            new Regex(@"^(.+?)(?: - )?\s?([｢「『【].+?[」｣』】](?:\s?MV)?.*)(?=\.\w{3,4})"),
            "%<1>s - %<2>s",
            "Reformat 'PERSON「TITLE」' followed by other info"),
        new(
            new Regex(@"(^.+) \[\s(.+)\s\]"),
            "%<1>s - %<2>s",
            "Reformat 'ARTIST [ TITLE ]'"),
        new(
            new Regex(@"^(.+)\s{1,}-\s{1,}['＂](.+)['＂]"),
            "%<1>s - %<2>s",
            """Reformat 'ARTIST - 'TITLE' ]', etc."""),
        new(
            new Regex(@"^(.+?)(?: - [｢「『【])(.+)(?:[」｣』】]).*(?=（Full Ver.）)"),
            "%<1>s - %<2>s",
            "Reformat 'ARTIST - \'TITLE\' ]'"),
        new(
            new Regex(@"(\d+) - \[(feat.+)\] (.+) ⧸ (.+)(?=\.\w{3,4})"),
            "%<4>s - %<4>s - %<1>s - %<3>s (%<2>s)"
        ),
        new(
            new Regex(@"(.+) ?⧸ ?(.+)(?= ：(?: \w+)?\.\w{3,4})"),
            "%<2>s - %<1>s"
        ),
        // Cleanup:
        new(
            new Regex(@" - - "),
            " - ",
            "Compress doubled hyphens"),
        new(
            new Regex(@" – "),
            " - ",
            "Replace en dashes with hyphens")
    };

    public static void Run(string workingDirectory, Printer printer)
    {
        Watch watch = new();

        DirectoryInfo dir = new(workingDirectory);

        var audioFilePaths = dir.EnumerateFiles("*.m4a");
        if (!audioFilePaths.Any())
        {
            printer.Warning("No audio files to rename were found.");
            return;
        }

        printer.Print($"Renaming {audioFilePaths.Count()} audio file(s)...");

        foreach (FileInfo filePath in audioFilePaths)
        {
            string newFileName = RenamePatterns.Aggregate(
                new StringBuilder(filePath.Name),
                (newFileNameSb, renamePattern) =>
                {
                    // Only continue if the current regex is a match.
                    var match = renamePattern.Regex.Match(newFileNameSb.ToString());
                    if (!match.Success)
                        return newFileNameSb;

                    // Delete the matched substring by index.
                    newFileNameSb.Remove(match.Index, match.Length);

                    // Work out the replacement text that should be inserted.
                    string insertText =
                        match.Groups.OfType<Group>()
                             .Select((gr, i) =>
                             (
                                SearchFor:   $"%<{i + 1}>s",
                                ReplaceWith: match.Groups[i + 1].Value
                             ))
                             .Aggregate(
                                 new StringBuilder(renamePattern.ReplaceWithPattern),
                                 (workingText, replacementParts) =>
                                     workingText.Replace(
                                        replacementParts.SearchFor,
                                        replacementParts.ReplaceWith),
                                 workingText => workingText.ToString()
                             );

                    // Insert the replacement text at the same starting position.
                    newFileNameSb.Insert(match.Index, insertText);

                    return newFileNameSb;
                },
                newFileNameSb => newFileNameSb.ToString());

            try
            {
                File.Move(
                    filePath.FullName,
                    Path.Combine(workingDirectory, newFileName));
                printer.Print($"• From: \"{filePath.Name}\"");
                printer.Print($"    To: \"{newFileName}\"");
            }
            catch (Exception ex)
            {
                printer.Error($"• Could not rename \"{filePath.Name}\": {ex.Message}");
            }
        }

        printer.Print($"Renaming done in {watch.ElapsedFriendly}.");
    }
}

File: src/CCVTAC.Main/PostProcessing/Renamer.fs
namespace CCVTAC.Main.PostProcessing

open CCVTAC.Main
open CCVTAC.Main.IoUtilities
open CCVTAC.Main.Settings.Settings
open CCFSharpUtils
open CCFSharpUtils.Collections
open CCFSharpUtils.Text
open System
open System.IO
open System.Text
open System.Text.RegularExpressions
open Startwatch.Library

module Renamer =

    let private toNormalizationForm (form: string) =
        match form.Trim().ToUpperInvariant() with
        | "D"  -> NormalizationForm.FormD
        | "KD" -> NormalizationForm.FormKD
        | "KC" -> NormalizationForm.FormKC
        | _    -> NormalizationForm.FormC

    let updateTextViaPattern isQuietMode (printer: Printer) (text: SB) (renamePattern: RenamePattern) : SB =
        let regex = Regex renamePattern.RegexPattern

        let matches = text.ToString() |> regex.Matches |> Rgx.successMatches |> Seq.rev |> Seq.toList

        let printSummary: unit =
            let patternDesc = if String.hasText renamePattern.Summary
                              then $"\"%s{renamePattern.Summary}\""
                              else $"`%s{renamePattern.RegexPattern}` (no description)"
            printer.Debug $"> Rename pattern %s{patternDesc} matched (%d{matches.Length}×)."

        /// Builds replacement text by substituting %<n>s placeholders with captured regex group values.
        /// Group values are trimmed, and indexing starts from 1 (because group 0 is the full match).
        let buildReplacementText (renamePattern: RenamePattern) (m: Match) : string =
            let sb = SB renamePattern.ReplaceWithPattern

            for i = 1 to m.Groups.Count - 1 do
                let placeholder = sprintf "%%<%d>s" i
                let value = if i < m.Groups.Count then m.Groups[i].Value.Trim() else String.Empty
                sb.Replace(placeholder, value) |> ignore

            sb.ToString()

        if List.isEmpty matches then
            text
        else
            if not isQuietMode then printSummary

            for m in matches do
                let replacementText = buildReplacementText renamePattern m
                text.Remove(m.Index, m.Length).Insert(m.Index, replacementText) |> ignore

            text

    let run userSettings workingDirectory (printer: Printer) : unit =
        let watch = Watch()
        let workingDirInfo = DirectoryInfo workingDirectory

        let audioFiles =
            workingDirInfo.EnumerateFiles()
            |> Seq.filter (fun f -> List.containsIgnoreCase f.Extension Files.audioFileExts)
            |> List.ofSeq

        match audioFiles with
        | [] ->
            printer.Warning "No audio files to rename were found."
        | _ ->
            printer.Debug $"""Renaming %s{String.fileLabelWithDesc "audio" audioFiles.Length}..."""

            for audioFile in audioFiles do
                let newFileName =
                    userSettings.RenamePatterns
                    |> List.fold
                        (fun (sb: SB) -> updateTextViaPattern userSettings.QuietMode printer sb)
                        (SB audioFile.Name)
                    |> _.ToString()

                try
                    let destinationPath =
                        Path.Combine(workingDirectory, newFileName)
                            .Normalize(toNormalizationForm userSettings.NormalizationForm)

                    File.Move(audioFile.FullName, destinationPath)

                    printer.Debug $"• From: \"%s{audioFile.Name}\""
                    printer.Debug $"    To: \"%s{newFileName}\""
                with exn ->
                    printer.Error $"• Error renaming \"%s{audioFile.Name}\": %s{exn.Message}"

            printer.Info $"Renaming done in %s{watch.ElapsedFriendly}."
